numismatic/feeds/base.py

```diff
@@ -261,55 +283,37 @@ def listen(self, symbol, channel=None, websocket_url=None):
                                     channel=channel,
                                     channel_info=channel_info, 
                                     client=self,
-                                    listener=None,
                                     handlers=self._get_handlers(),
                                     )
-        # FIXME: find a better way to do this
-        subscription.listener = self._listener(subscription)
-        subscription.raw_stream.sink(
-            partial(self.__handle_packet, subscription=subscription))
+        self.subscriptions.append(subscription)
+        asyncio.ensure_future(subscription.start())
         return subscription
 
-    async def _listener(self, subscription):
-        if self.websocket is None:
-            self.websocket = await self._connect(subscription)
-        await self._subscribe(subscription)
+    async def _listener(self):
+        await self._connect()
```

Comment:
Regarding connect, I have an idea. Lets create a connection class within the webocket_client class. We will implement the dunders `__enter__` and `__exit__`, and then we can do this:
```
with self.connection():
  while True:
    ...
```
Even the `ConnectionClosed` exception can be handled well..